uxarray/helpers.py

```diff
@@ -556,3 +556,111 @@ def _normalize_in_place(node):
 
     return list(np.array(node) / np.linalg.norm(np.array(node), ord=2))
+
+
+@njit
+def _pt_within_gcr(pt_cart, gcr_end_pt_cart_1, gcr_end_pt_cart_2):
+    """Helper function to determine if a point lies within the interval of the great circle arc, the accuracy is
+    within 1.0e-12 The great circle arc will be represented as two ndarrays: gcr_end_pt_cart_1 and gcr_end_pt_cart_2
+    since numba lacks support for the nested array
+
+    Parameters
+    ----------
+    pt_cart : float ndArray
+        3D Cartesian Coordinates [x, y, z] for the input point
+    gcr_end_pt_cart_1 : float ndArray
+        3D Cartesian Coordinates list [x1, y1, z1], one of the end points of the great circle arc
+    gcr_end_pt_cart_2 : float ndArray
+    3D Cartesian Coordinates list [x2, y2, z2], another end points of the great circle arc
+
+    Returns
+    -------
+    bool
+        True if the point lies within the interval of the gcr, False otherwise.
+
+    """
+    cross = np.cross(gcr_end_pt_cart_1, gcr_end_pt_cart_2)
+    dot = cross[0] * pt_cart[0] + cross[1] * pt_cart[1] + cross[2] * pt_cart[2]
+    if np.absolute(dot - 0) > 1.0e-12:
+        return False
+
+    # If we have determined the point lies on the gcr plane, we only need to check if the pt's longitude lie within
+    # the gcr
+    pt_lonlat_rad = _convert_node_xyz_to_lonlat_rad(list(pt_cart))
+    gcr_lonlat_rad_1 = _convert_node_xyz_to_lonlat_rad(list(gcr_end_pt_cart_1))
+    gcr_lonlat_rad_2 = _convert_node_xyz_to_lonlat_rad(list(gcr_end_pt_cart_2))
+
+    # Special case: when the gcr and the point are all on the same longitude line:
+    if gcr_lonlat_rad_1[0] == gcr_lonlat_rad_2[0] == pt_lonlat_rad[0]:
+        # Now use the latitude to determine if the pt falls between the interval
+        return within(gcr_lonlat_rad_1[1], pt_lonlat_rad[1], gcr_lonlat_rad_2[1])
+
+    # First we need to deal with the longitude wrap-around case
+    # x0--> 0 lon --> x1
+    if np.absolute(gcr_lonlat_rad_2[0] - gcr_lonlat_rad_1[0]) >= np.deg2rad(180):
+        if within(np.deg2rad(180), gcr_lonlat_rad_1[0], np.deg2rad(360)) and within(0, gcr_lonlat_rad_2[0],
+                                                                                    np.deg2rad(180)):
+            return within(gcr_lonlat_rad_1[0], pt_lonlat_rad[0], np.deg2rad(360)) or within(0, pt_lonlat_rad[0],
+                                                                                            gcr_lonlat_rad_2[0])
+        elif within(np.deg2rad(180), gcr_lonlat_rad_2[0], np.deg2rad(360)) and within(0, gcr_lonlat_rad_1[0],
+                                                                                      np.deg2rad(180)):
+            # x1 <-- 0 lon <-- x0
+            return within(gcr_lonlat_rad_2[0], pt_lonlat_rad[0], np.deg2rad(360)) or within(0, pt_lonlat_rad[0],
+                                                                                            gcr_lonlat_rad_1[0])
```

Comment:
Can use `np.abs()` instead of `np.absolute()` for better readability

Reply:
> Use `np.pi` or `2*np.pi` to reduce unnecessary function calls with `np.deg2rad()`

`np.deg2rad` will have better readability here and will avoid some hand-written bugs compared to using `np.pi`. Does it have a significantly lower performance?

Reply:
Since the function is operates on a single point, it would call `np.deg2rad()` 12 times for each point that it's called on. 

What about defining `np.pi` and `2*np.pi` as constants as naming them according to the context of the problem.

Reply:
> Since the function is operates on a single point, it would call `np.deg2rad()` 12 times for each point that it's called on.
> 
> What about defining `np.pi` and `2*np.pi` as constants as naming them according to the context of the problem.

Can you explain a bit more why calling `np.deg2rad()` will be way slower than using the `2*np.pi` calculation here?

Since the entire calculation involves lots of calculations, I want to reduce the "wheel rebuilding" as much as possible to eliminate all unnecessary risks.

Reply:
I'm not certain that it would be _**way slower**_ than using a constant, I would need to run some tests and look into the source code of `np.deg2rad()`. But since your function is written to operate on each individual point, for a large dataset we would have potentially thousands or more of duplicate function calls. 

For a vectorized implementation (i.e. one function call for the set of all points), I wouldn't be opposed to using `np.deg2rad()`, but for this implementation it makes more sense to use a constant.

Reply:
> I'm not certain that it would be _**way slower**_ than using a constant, I would need to run some tests and look into the source code of `np.deg2rad()`. But since your function is written to operate on each individual point, for a large dataset we would have potentially thousands or more of duplicate function calls to `np.deg2rad()`
> 
> For a vectorized implementation (i.e. one function call for the set of all points), I wouldn't be opposed to using `np.deg2rad()`, but for this implementation it makes more sense to use a constant.

I still don't quite get it here. So basically what's inside the `np.deg2rad()` will be the calculation using the constant, so why it will be different than using the `2*np.pi`?

If we will keep using the constant `2*np.pi` here, we will have lots of duplication codes here, which can be turned into a function call to our "handwritten degree to radian conversion function", which will make no difference from calling the `np.deg2rad`

I think the final goal is still vectorizing every function as much as possible.

Reply:
The difference is when you call the function, it's still converting from degrees to radians (deg2rad(x) is x * pi / 180). If we define it as a constant **outside of the scope** of the function, we can avoid needing to recalculate it every time we call it.

Reply:
> The difference is when you call the function, it's still converting from degrees to radians (deg2rad(x) is x * pi / 180). If we define it as a constant **outside of the scope** of the function, we can avoid needing to recalculate it every time we call it.

Got it. So you just mean using the constant for the `deg2rad(180)` and `deg2rad(360)` here, while still using the `np.deg2rad()` for other dynamic points

Reply:
Exactly!